### PR TITLE
Add auth/kubernetes

### PR DESCRIPTION
**Pull request by johanot**
Implements auth via Kubernetes serviceaccount (https://www.vaultproject.io/api-docs/auth/kubernetes)

Also: Added basic tests for all endpoints, except login. The reason being that it seems to require a dockertest server with a webserver that can simulate a TokenReviewResponse from a k8s apiserver. I think adding something like "httpmock" to you dockertest-server repository could do the trick? Perhaps, except that Vault requires TLS-connections to the APIserver, so it would require a TLS-proxy container as well, or a specialized docker image for handling all of the above.

tldr, it was too much work for me to find a fitting solution for this at the moment. I would like to request help/suggestions for implementing the login test later at some point.

### CODE REVIEW

**Comment by jmgilman** [comment]
Thanks for your contribution! I updated some dependencies, would you mind updating the PR to resolve the conflict. Also, I'm not very familiar with k8s integration in Vault, but looking through the code I'm wondering what the login test is currently doing? On the surface, it looks sufficient, but I'm not sure if there are some underlying mechanisms being missed.

For context, the reason a mocking library wasn't used was due to the fact that there's very little application logic in this crate and most of it revolves around correctly using the Vault API. Thus, it was deemed more beneficial to bring up a local container to actually test that requests/responses were being generated/parsed correctly. 

That being said, the desire is to avoid testing Vault unnecessarily (we trust Hashicorp is doing a great job there) and only test our consumption of their API. If a test becomes unnecessarily complex, it's perfectly fine to send a request and expect a specific failure back (i.e. Vault complains the endpoint isn't TLS). Again, the point is more testing the quality of the requests in this instance than that Vault is doing what it's supposed to be doing.

Hopefull that makes sense!

**Comment by johanot** [comment]
> would you mind updating the PR to resolve the conflict

Sure.

> Also, I'm not very familiar with k8s integration in Vault.

Summarized: Containers in Kubernetes can optionally get a service-account token (JWT) injected, which can be used for authentication against the Kubernetes control-plane. The JWT contains information like namespace and name, which Vault uses for binding the Kubernetes token to a role internally. Using the TokenReview endpoint in the Kubernetes apiserver, vault-server will - upon login - try to contact the k8s apiserver to verify that the provided token is genuine and if so, match the token claims against a role and ultimately assign vault-policies to it.

> If a test becomes unnecessarily complex, it's perfectly fine to send a request and expect a specific failure back (i.e. Vault complains the endpoint isn't TLS).

Ok. I can enable the login test and skip setting up any k8s apiserver mock entirely? In that case, vault will fail login with something like "permission denied" (from the client perspective), and the Vault server log might contain something like "login unauthorized due to: connection refused" or whatever it says when a tcp-port isn't open. The problem of course being that asserting the response to be "permission denied" will not make a very robust login test-case.

ref: https://github.com/hashicorp/vault-plugin-auth-kubernetes/blob/main/path_login.go#L105-L110

@jmgilman Thank you for your feedback. I will revisit this PR on Monday and see what I can do.

**Comment by jmgilman** [comment]
> Ok. I can enable the login test and skip setting up any k8s apiserver mock entirely? In that case, vault will fail login with something like "permission denied" (from the client perspective), and the Vault server log might contain something like "login unauthorized due to: connection refused" or whatever it says when a tcp-port isn't open. The problem of course being that asserting the response to be "permission denied" will not make a very robust login test-case.

Yeah, that's the trade-off with just looking for a specific error condition rather than fully testing a successful scenario. Balancing complexity with the reliability of the test is what I've typically done. You can find an example of a more complex test case in the [OIDC test](https://github.com/jmgilman/vaultrs/blob/master/vaultrs-login/tests/login.rs#L118) for the login crate. In that test, I opted to simulate the whole OIDC workflow since an accurate login implementation is important to test the correctness of the login crate in particular. 

I'm more than fine opting for the less complex approach now and potentially re-evaluating in the future. It looks like the most complete way is going to involve either finding or building a mock k8s authentication container and then running it with `docker-test`.

**Comment by johanot** [comment]
@jmgilman Ok, added a full integration test of kubernetes login by use of the nginx testserver ( https://github.com/jmgilman/dockertest-server/pull/9 ). The test adds a mock-response to the nginx server with TLS-enabled to satisfy Vault kubernetes login checks.

Remember to change the pin to my fork here:

https://github.com/jmgilman/vaultrs/pull/18/commits/7633161975fc25e554044487ee5785b744644735#diff-2e9d962a08321605940b5a657135052fbcef87b5e360662bb527c96d9a615542R39

, when dockertest-server#9 is merged.

**Comment by johanot** [comment]
@jmgilman Any chance you'll get time for helping this getting merged? We are using this in production now, so it's a bit sad for us that we need to maintain a fork. Let me know if I can do anything to help on the project in general.

**Comment by jmgilman** [comment]
@johanot Sorry this fell off my radar again. Can you please run `cargo fmt`?

**Comment by johanot** [comment]
@jmgilman fmt done :) cfc8571

**Comment by johanot** [comment]
@jmgilman Thanks for merge! Do you want me to update the README or something else before next release?

**Comment by jmgilman** [comment]
No, I'm going to do the release today, just working on updating dependencies. Updating `hmac` to v0.12.1 broke some of the tests you added, just digging into it now. You're welcome to see if you can find the issue.

**Comment by jmgilman** [comment]
Ok, figured it out. All changes should be available in `v0.6.0`.
